Add order book documentation (#85)
zeitgeistpm/documentation: docs/learn/order-book.md

Reviewer: I think we should mention the creator fees.

Reviewer: We may also want to mention that the order book functions using order hashes.

Reviewer: Have you settled on one particular spelling of _order book_?

market's base asset (collateral). The order book can be set as the market's
scoring rule. The operations within the order book are primarily managed through
three fundamental functions: `place_order`, `remove_order`, and `fill_order`.

### Place Order

This function is used to introduce a new order into the market. When traders
decide to buy or sell an asset, they use `place_order` to specify the details of
their order, the quantity they wish to buy the `taker_asset` for `taker_amount`
and sell the `maker_asset` for `maker_amount`. This order is then added to the
order book, where it waits to be matched by taker(s).

It is important to mention that the users of the Zeitgeist orderbook do specify
the price of the assets implicitly. This means that the `place_order` does not
take a price as an argument, but rather the `taker_amount` and `maker_amount`.
This means the caller of the `place_order` function does willingly sell the
`maker_amount` of the `maker_asset` to buy the given `taker_amount` of the
`taker_asset`. After the execution of `place_order` the `maker_amount` of the
`maker_asset` is reserved for the order. This means it is not available for the

> user to use in other transactions.

Reviewer: Would be nice to have a formula for the price here.

Author: 733c3bd187ad018a0b31588b080e4b38b9156ad3

### Fill Order

The crux of any trading activity is the completion of trades, which is
facilitated by the `fill_order` function. This function is triggered when a

> taker matches a given maker order. The fill_order function will execute the

Reviewer: ```suggestion
taker matches a given maker order. The `fill_order` function will execute the
```

trade, transferring the specified asset between parties at the agreed price, and
update the order book to reflect the completion of the trade.

> The taker has the ability to partially fill the order for a given

Reviewer: I'm confused by this description. I think you're confusing maker and taker here. The taker is the one who fills the order, so I'm not quite sure what "This amount specifies the partial amount of what the maker wants to fill." is referring to.

Reviewer: What I'm saying is that the sentence "... the maker wants to fill" doesn't really work because the maker doesn't fill orders.

Author: Then it's `the maker wants to get filled`

Author: But anyways, I deleted it. Can you approve?

Reviewer: The next sentence also mentions the maker filling an order. ;)

Author: c687c1dfd81d3665994587383231d98088cb782b

`maker_partial_fill` amount. This amount specifies the partial amount of what
the maker wants to fill. The maker always wants to fill the `taker_amount` of

> the (in `place_order`) specified taker asset. Thus, the `maker_partial_fill` (if

Reviewer: ```suggestion
the specified taker asset (in `place_order`). Thus, the `maker_partial_fill` (if

This means the caller of the `place_order` function does willingly sell the
`maker_amount` of the `maker_asset` to buy the given `taker_amount` of the
`taker_asset`. After the execution of `place_order` the `maker_amount` of the
`maker_asset` is reserved for the order. This means it is not available for the
user to use in other transactions.

##### Price Example

Let's assume you wish to purchase `42` outcome tokens, referred to as
`taker_asset`. You decide to offer `7` ZTG, your `maker_asset`, for this
transaction. Essentially, this means you're ready to exchange `7` ZTG for `42`
outcome tokens. By doing so, you've implicitly set a rate of `6` outcome tokens
per `1` ZTG (since `42` divided by `7` equals `6`). Therefore, if a taker agrees
to your proposed rate, they need to supply `42` outcome tokens (or fewer, in the
case of partial order fulfillment) and, in return, will get `7` ZTG from you (or
less, depending on the partial fulfillment scenario).

It's important to note that the market creator's fee is always deducted in the
market's base currency, which in our example is ZTG. So, if the market creator

> charges a `1%` fee, the taker ends up receiving `6.3` ZTG, while the market
> creator gets `0.7` ZTG. However, the `taker_amount` you receive as the maker

Reviewer: ```suggestion
charges a `1%` fee, the taker ends up receiving `6.93` ZTG, while the market
creator gets `0.07` ZTG. However, the `taker_amount` you receive as the maker
```
